File: sandboxes/npm/analyze.js
#!/usr/bin/env node

const { spawnSync } = require('child_process');
const { argv } = require('process');

console.log(argv);
const nodeBin = argv.shift();
const scriptPath = argv.shift();
console.log(argv);

if (argv.length < 2 || argv > 4) {
  console.log(`Usage: ${nodeBin} ${scriptPath} [--local file | --version version] phase pkg`);
  process.exit(1);
}

var localFile = null;
var ver = null;
switch (argv[0]) {
  case '--local':
    argv.shift();
    localFile = argv.shift();
    break;
  case '--version':
    argv.shift();
    ver = argv.shift();
    break;
}

const phase = argv.shift();
const pkg = argv.shift();

if (phase != 'all') {
  console.log('Only "all" phase is supported at the moment.');
  process.exit(1);
}

// Specify the package to install.
const installPkg = localFile ? localFile : (ver ? `${pkg}@${ver}` : pkg);

let result = spawnSync('npm', ['init', '--force'], { stdio: 'inherit' });
if (result.status != 0) {
  throw 'Failed to init npm';
}

result = spawnSync('npm', ['install', installPkg], { stdio: 'pipe', encoding: 'utf8' });
console.log(result.stdout + result.stderr);

if (result.status == 0) {
  console.log('Install succeeded.');
} else {
  if (result.stderr.includes('is not in the npm registry.')) {
    process.exit(0);
  }
  console.log('Install failed.');
  process.exit(1);
}

try {
  require(pkg);
} catch (e) {
  console.log(`Failed to import ${pkg}: ${e}`);
}

File: sandboxes/rubygems/analyze.rb
#!/usr/bin/env ruby
#
require 'find'
require 'open3'
require 'pathname'

def install(package, version, local_file)
  cmd = ["gem", "install"]
  if local_file
    cmd << local_file
  else
    if version
      cmd << "-v"
      cmd << version
    end
    cmd << package
  end

  output, status = Open3.capture2e(*cmd)
  puts output

  if status.success?
    puts "Install succeeded."
    return
  end

  puts "Install failed."
  if output.include? "Could not find a valid gem"
    exit 0
  end

  exit 1
end

if ARGV.length < 2 || ARGV.length > 4
  puts "Usage: #{$0} [--local file | --version version] phase package"
  exit 1
end

local_file = nil
version = nil

case ARGV[0]
when "--local"
  ARGV.shift
  local_file = ARGV.shift
when "--version"
  ARGV.shift
  version = ARGV.shift
end

phase = ARGV.shift
package = ARGV.shift

if phase != "all"
  puts "Only \"all\" phase is supported at the moment"
  exit 1
end

install(package, version, local_file)
spec = Gem::Specification.find_by_name(package)

spec.require_paths.each do |require_path|
  if Pathname.new(require_path).absolute?
    lib_path = Pathname.new(require_path)
  else
    lib_path = Pathname.new(File.join(spec.full_gem_path, require_path))
  end

  Find.find(lib_path.to_s) do |path|
    if path.end_with?('.rb')
      relative_path = Pathname.new(path).relative_path_from(lib_path)

      require_path = relative_path.to_s.delete_suffix('.rb')
      puts "Loading #{require_path}"
      begin
        require require_path
      rescue Exception => e
        puts "Failed to load #{require_path}: #{e}"
      end
    end
  end
end
